src/validationinterface.cpp

```diff
@@ -185,6 +215,151 @@ void ThreadNotifyWallets(CBlockIndex *pindexLastTip)
+        //   - For each transaction that became conflicted by this block:
+        //     - 4️⃣ Trial-decrypt the transaction's shielded outputs.
+        //     - If the transaction belongs to the wallet:
+        //       - 5️⃣ Add or update the transaction, and mark it as dirty.
+        //   - For each transaction in the block:
+        //     - 6️⃣ Trial-decrypt the transaction's shielded outputs.
+        //     - If the transaction belongs to the wallet:
+        //       - 7️⃣ Add or update the transaction, and mark it as dirty.
+        //   - Update the wallet's view of the chain tip.
+        //     - 8️⃣ In `zcashd`, this is when we increment note witnesses.
+        // - For each transaction in the mempool:
+        //   - 9️⃣ Trial-decrypt the transaction's shielded outputs.
+        //   - If the transaction belongs to the wallet:
+        //     - 🅰️ Add or update the transaction, and mark it as dirty.
+        //
+        // Steps 2️⃣, 3️⃣, 5️⃣, 7️⃣, 8️⃣, and 🅰️ are where wallet state is updated,
+        // and the relative order of these updates must be preserved in order to
+        // avoid breaking any internal assumptions that the wallet makes.
+        //
+        // Steps 1️⃣, 4️⃣, 6️⃣, and 9️⃣ can be performed at any time, as long as
+        // their results are available when their respective conditionals are
+        // evaluated. We therefore refactor the above workflow to enable the
+        // trial-decryption work to be batched and parallelised:
+        //
+        // - For each block in the old chain:
+        //   - For each transaction in the block:
+        //     - Accumulate its Sprout, Sapling, and Orchard outputs.
+        // - For each block in the new chain:
+        //   - For each transaction that became conflicted by this block:
+        //     - Accumulate its Sprout, Sapling, and Orchard outputs.
+        //   - For each transaction in the block:
+        //     - Accumulate its Sprout, Sapling, and Orchard outputs.
+        //
+        // - 1️⃣4️⃣6️⃣9️⃣ Trial-decrypt the Sprout, Sapling, and Orchard outputs.
+        //   - This can split up and batch the work however is most efficient.
+        //
+        // - For each block in the old chain, from its tip to the fork point:
+        //   - For each transaction in the block:
+        //     - If the transaction has decrypted outputs, or transparent inputs
+        //       that belong to the wallet:
+        //       - 2️⃣ Add or update the transaction, and mark it as dirty.
+        //   - Update the wallet's view of the chain tip.
+        //     - 3️⃣ In `zcashd`, this is when we decrement note witnesses.
+        // - For each block in the new chain, from the fork point to its tip:
+        //   - For each transaction that became conflicted by this block:
+        //     - If the transaction has decrypted outputs, or transparent inputs
+        //       that belong to the wallet:
```

Comment:
What about notes belonging to this wallet that become spent? Also what about transparent outputs belonging to the wallet?